reference/src/main/scala/akka/stream/alpakka/reference/model.scala

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright (C) 2016-2018 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.stream.alpakka.reference
+
+// rename Java imports if the name clashes with the Scala name
+import java.lang.{Long => JavaLong}
+import java.util.{List => JavaList, Map => JavaMap}
+
+import akka.util.ByteString
+
+import scala.annotation.varargs
+import scala.collection.immutable
+import akka.japi.Pair
+
+import scala.util.{Failure, Success, Try}
+
+/**
+ * Use "Read" in message data types to signify that the message was read from outside.
+ */
+final class ReferenceReadMessage private (
+    val data: immutable.Seq[ByteString] = immutable.Seq.empty,
+    val bytesRead: Try[Int] = Success(0)
+) {
+  def withData(data: immutable.Seq[ByteString]): ReferenceReadMessage =
+    copy(data = data)
+
+  def withBytesRead(bytesRead: Try[Int]): ReferenceReadMessage =
+    copy(bytesRead = bytesRead)
+
+  /**
+   * Java API
+   *
+   * If the model class is meant to be also consumed from the user API,
+   * but the attribute class is Scala specific, create getter for Java API.
+   */
+  def getData(): JavaList[ByteString] = {
+    import scala.collection.JavaConverters._
+    data.asJava
+  }
+
+  /**
+   * Java API
+   *
+   * If the model class is scala.util.Try, then two getters should be created.
+   * One for getting the value, and another for getting the exception.
+   *
+   * Should return null if the Try contains a Failure.
+   */
+  def getBytesRead(): Integer = bytesRead match {
+    case Success(bytesRead) => bytesRead
+    case Failure(_) => null
+  }
+
+  /**
+   * Java API
+   *
+   * Return the exception if the Try contains a Failure, otherwise return a null.
+   */
+  def getBytesReadFailure(): Throwable = bytesRead match {
+    case Success(_) => null
+    case Failure(ex) => ex
+  }
+
+  private def copy(data: immutable.Seq[ByteString] = data, bytesRead: Try[Int] = bytesRead) =
+    new ReferenceReadMessage(data, bytesRead)
+}
+
+object ReferenceReadMessage {
+  def apply(): ReferenceReadMessage = new ReferenceReadMessage()
```

Comment:
`create` for Java? 
Even if this is constructed internally only, for testing it should be made available.

Reply:
Good point.